### PR TITLE
Lpmem hashtable

### PR DESCRIPTION
**Pull request by eshiple**
implemented hash table block allocator + lblk->pblk translation with simd, fixed json output for filebench, cuckoo hashtable block allocator,

### CODE REVIEW

**Comment by iangneal** [comment]
Too many files added! Remove the benchout stuff at least.

File: bench/BenchmarkProcesses.py [hunk 1]
```diff
@@ -53,9 +53,9 @@ def start(self):
         subprocess.run(mkdir_args, check=True, stdout=DEVNULL, stderr=DEVNULL)
         '''
 
-        mkfs_args = [ str(self.kernfs_path / 'mkfs.sh') ]
-        proc = subprocess.run(mkfs_args, cwd=self.kernfs_path, check=True,
-                              stdout=DEVNULL, stderr=DEVNULL)
+        mkfs_args = [ 'sudo', '-E', str(self.kernfs_path / 'mkfs.sh') ]
```
Comment (iangneal) [inline bench/BenchmarkProcesses.py:56]: Do we need this?

File: kernfs/extents.c [hunk 2]
```diff
@@ -1303,7 +1313,7 @@ static int mlfs_ext_grow_indepth(handle_t *handle,
 	if (ext_depth(handle, inode))
 		goal = mlfs_idx_pblock(EXT_FIRST_INDEX(ext_inode_hdr(handle, inode)));
 	goal = mlfs_inode_to_goal_block(inode);
-
+	//printf("hello world\n");
```
Comment (iangneal) [inline kernfs/extents.c]: I don't think we need this :)

File: kernfs/extents.c [hunk 3]
```diff
@@ -2792,6 +2802,89 @@ static mlfs_lblk_t mlfs_ext_determine_hole(handle_t *handle, struct inode *inode
...
+		tsc_start = asm_rdtscp();
+        start_cache_stats();
+    }
+#endif
+
+    assert(map_arr->m_len <= 8 && "fs_get_blocks m_len > 8");
+    struct super_block *sblk = sb[g_root_dev];
+    //printf("retrieving %u blocks\nmap_arr: ", map_arr->m_len);
+	int create = flags & MLFS_GET_BLOCKS_CREATE_DATA;
+	int create2 = flags & MLFS_GET_BLOCKS_CREATE_META;
```
Comment (iangneal) [inline kernfs/extents.c]: create2 isn't a great variable name

File: kernfs/extents.c [hunk 4]
```diff
@@ -3185,6 +3280,32 @@ int mlfs_ext_get_blocks(handle_t *handle, struct inode *inode,
 int mlfs_ext_truncate(handle_t *handle, struct inode *inode,
 		mlfs_lblk_t start, mlfs_lblk_t end)
 {
+
+	printf("HELLO WORLD\n");
```
Comment (iangneal) [inline kernfs/extents.c]: plz remove

File: kernfs/fs.c [hunk 5]
```diff
@@ -195,12 +201,16 @@ void show_kernfs_stats(void)
...
         json_object_object_add(root, "search", search);
     }
+	json_object_array_add(kernfs_stats_json, root);
 
     // Convert JSON to a string
-	const char *js_str = json_object_get_string(root);
-
+	const char *js_str = json_object_get_string(kernfs_stats_json);
+	//const char *js_str = json_object_get_string(root);
+	printf("%s\n", js_str);
```
Comment (iangneal) [inline kernfs/fs.c]: Extra

File: kernfs/lpmem_ghash.c [hunk 6]
```diff
@@ -0,0 +1,1620 @@
...
+{
+// #ifndef SIMPLE_ENTRIES
+//    ent->size = new_range;
+//    nvm_update(pop, node_index);
+// #endif
+}
+// Very similar to lookup, but we also update the entry at the end.
+int pmem_nvm_hash_table_update(paddr_t         key,
+                          size_t          new_range) {
+#ifdef SEQ_STEP
```
Comment (iangneal) [inline kernfs/lpmem_ghash.c]: In general maybe just remove things that aren't being used to make this easier to read

File: kernfs/migrate.c [hunk 7]
```diff
@@ -374,7 +419,7 @@ int migrate_blocks(uint8_t from_dev, uint8_t to_dev, isolated_list_t *migrate_li
 			continue;
 		}
 
-		mlfs_assert(map.m_pblk != 0);
+		// mlfs_assert(map.m_pblk != 0);
```
Comment (iangneal) [inline kernfs/migrate.c]: Can we add back this assert for each case?

File: kernfs/tests/Makefile [hunk 8]
```diff
@@ -20,16 +21,17 @@ API_DIR := $(abspath ../../../nvm-file-indexing/api)
...
+	-Wl,-rpath=$(abspath $(API_DIR)/build) -L$(abspath $(API_DIR)/build) -lNVMIndexingAPI -lm
 LD_FLAGS_CXX = -lpthread -lm -L$(NVML_DIR)/nondebug/ -lpmem -lrt \
 	-Wl,-rpath=$(abspath $(GLIB_DIR)/glib/.libs) -L$(abspath $(GLIB_DIR)/glib/.libs) \
 	-Wl,-rpath=$(abspath $(API_DIR)/build) -L$(abspath $(API_DIR)/build) -lNVMIndexingAPI
 
-DEBUG = -g -O0
+#DEBUG = -g -O0
 #DEBUG = -g -Ofast
-#DEBUG = -g -O3
+DEBUG = -g -O3 -march=skylake-avx512 -mavx -mavx512f -mavx512vl -mavx512cd -mavx512bw -mavx512dq
```
Comment (iangneal) [inline kernfs/tests/Makefile]: DEBUG should have -O0 since it's easier to step through

File: libfs/src/filesystem/dirent.c [hunk 9]
```diff
@@ -102,11 +102,27 @@ uint8_t *get_dirent_block(struct inode *dir_inode, offset_t offset)
...
         if (enable_perf_stats)
             tsc_begin = asm_rdtscp();
-		ret = bmap(dir_inode, &bmap_req);
+		int blk_count_found = 0, block_num = 0;;
+		if(IDXAPI_IS_HASHFS()) {
+			ret = bmap_hashfs(dir_inode, &bmap_req_arr);
+			blk_count_found = bmap_req_arr.blk_count_found;
+			block_num = bmap_req_arr.block_no[0];
+		}
+		else {
```
Comment (iangneal) [inline libfs/src/filesystem/dirent.c]: I personally prefer the `} else {` style

File: libfs/src/filesystem/dirent.c [hunk 10]
```diff
@@ -115,23 +131,24 @@ uint8_t *get_dirent_block(struct inode *dir_inode, offset_t offset)
...
 
 			mlfs_io_wait(dir_inode->dev, 1);
-
+	
 			d_block = dcache_alloc_add(dev, dir_inode->inum,
-					offset, data, bmap_req.block_no, (struct fs_log*)g_fs_log);
-
+				offset, data, block_num, (struct fs_log*)g_fs_log);
+			
+
```
Comment (iangneal) [inline libfs/src/filesystem/dirent.c]: Remove extra lines

File: libfs/src/mkfs/mkfs.c [hunk 11]
```diff
@@ -526,7 +547,29 @@ void iappend(uint8_t dev, uint32_t inum, void *xp, int n)
...
+					// addr_t ndb = ondisk_sb.ndatablocks;
+					// uint64_t ent_num_bytes = sizeof(paddr_t) * ndb;
+  					// int ent_num_blocks_needed = 1 + ent_num_bytes / g_block_size_bytes;
+  					// if(ent_num_bytes % g_block_size_bytes != 0) {
+    				// 	++ent_num_blocks_needed;
+  					// }
+					// addr_t hash = key;
+					// addr_t modded = hash % (ndb - ent_num_blocks_needed);
+					// addr_t pblk = modded + ent_num_blocks_needed;
+					addr_t index;
```
Comment (iangneal) [inline libfs/src/mkfs/mkfs.c:540]: Are all the comments needed?

File: libfs/src/mkfs/mkfs.c [hunk 12]
```diff
@@ -526,7 +547,29 @@ void iappend(uint8_t dev, uint32_t inum, void *xp, int n)
...
+					if(IDXAPI_IS_CUCKOOFS()) {
+						//getchar();
+						pmem_cuckoohash_create(inum, 0, &index);
```
Comment (iangneal) [inline libfs/src/mkfs/mkfs.c]: Also are these tabs or spaces?

File: libfs/src/storage/storage.h [hunk 13]
```diff
@@ -41,7 +41,7 @@ extern "C" {
 */
 
 // device size in bytes
-static uint64_t dev_size[g_n_devices + 1] = {0UL, 21137195008UL, 0UL, 0UL, 2111832064UL};
+static uint64_t dev_size[g_n_devices + 1] = {0UL, 10567548928UL, 0UL, 0UL, 2111832064UL};
```
Comment (iangneal) [inline libfs/src/storage/storage.h:44]: Don't stage these changes generally

File: libfs/src/storage/storage.c [hunk 14]
```diff
@@ -5,11 +5,11 @@ struct block_device *g_bdev[g_n_devices + 1];
 #if 1
 char *g_dev_path[] = {
 	(char *)"unused",
-	(char *)"/dev/dax0.2",
+	(char *)"/dev/dax0.4",
 	//(char *)"/tmp/devone",
 	(char *)"PCIe-SSD",
 	(char *)"/backup/mlfs_hdd",
-	(char *)"/dev/dax0.1",
+	(char *)"/dev/dax0.2",
```
Comment (iangneal) [inline libfs/src/storage/storage.c:12]: Don't stage these changes generally.

File: libfs/tests/write_bench.c [hunk 15]
```diff
@@ -0,0 +1,156 @@
+#include <stdio.h>
```
Comment (iangneal) [inline libfs/tests/write_bench.c]: Can you add an explanation at the top about why this exists and how you use it?

File: libfs/tests/read_bench.c [hunk 16]
```diff
@@ -0,0 +1,169 @@
+#include <stdio.h>
```
Comment (iangneal) [inline libfs/tests/read_bench.c]: Can you add an explanation at the top about why this test exists and how you use it?

File: utils/use_dax.sh [hunk 17]
```diff
@@ -8,7 +8,7 @@ fi
 if [[ $1 == "bind" ]];
 then
 	sudo ndctl create-namespace -m dax -e namespace0.0 -f
-	sudo ndctl create-namespace -m dax -e namespace1.0 -f
+	sudo ndctl create-namespace -m dax -e namespace0.1 -f
```
Comment (iangneal) [inline utils/use_dax.sh:11]: Don't think we need this either

**Review by iangneal — CHANGES_REQUESTED** [review]
Left you some general comments as well.

File: bench/BenchmarkProcesses.py (hunk 1 above)
Comment (eshiple) [reply bench/BenchmarkProcesses.py:56]: I think this was our solution to adding propogating env through bash scripts calling other bash scripts

File: kernfs/migrate.c (hunk 7 above)
Comment (eshiple) [reply kernfs/migrate.c]: i think so

File: kernfs/tests/mkfs.sh [hunk 18]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/sudo /bin/bash
+#! /bin/bash
```
Comment (eshiple) [inline kernfs/tests/mkfs.sh:1]: here is where we removed sudo and added it to the python script

File: libfs/src/mkfs/mkfs.c (hunk 12 above)
Comment (eshiple) [reply libfs/src/mkfs/mkfs.c]: tabs?

File: bench/BenchmarkProcesses.py (hunk 1 above)
Comment (iangneal) [reply bench/BenchmarkProcesses.py:56]: I don't think we need the sudo though if you `sudo chown user:user /dev/daxWhatever`

File: kernfs/hash_functions.h [hunk 19]
```diff
@@ -0,0 +1 @@
+../../nvm-file-indexing/api/common/hash_functions.h
```
Comment (iangneal) [inline kernfs/hash_functions.h]: Might want to copy this rather thank sim-linking.

File: kernfs/lookup3.c [hunk 20]
```diff
@@ -0,0 +1,375 @@
+/*
```
Comment (iangneal) [inline kernfs/lookup3.c]: Do we need this file?

File: kernfs/lpmem_cuckoo_hash_impl.c [hunk 21]
```diff
@@ -0,0 +1,373 @@
+/*
```
Comment (iangneal) [inline kernfs/lpmem_cuckoo_hash_impl.c]: Do we need this file?

File: kernfs/lpmem_cuckoo_hash_impl.h [hunk 22]
```diff
@@ -0,0 +1,197 @@
+/*
```
Comment (iangneal) [inline kernfs/lpmem_cuckoo_hash_impl.h]: Do we need this file?

File: kernfs/lpmem_cuckoohash.c [hunk 23]
```diff
@@ -0,0 +1,123 @@
+#include <stdbool.h>
```
Comment (iangneal) [inline kernfs/lpmem_cuckoohash.c]: Do we need this file?

File: kernfs/lpmem_cuckoohash.h [hunk 24]
```diff
@@ -0,0 +1,48 @@
+#ifndef __NVM_IDX_CUCKOO_HASH__
```
Comment (iangneal) [inline kernfs/lpmem_cuckoohash.h]: Do we need this file?

File: kernfs/lpmem_ghash.c [hunk 25]
```diff
@@ -0,0 +1,883 @@
...
+  
+}
+
+#pragma GCC pop_options
+
+/*
+ * Send help.
+ *
+ * Also I think we can only use this in libfs.
+ */
```
Comment (iangneal) [inline kernfs/lpmem_ghash.c:358]: This comment can definitely be removed.

File: kernfs/lpmem_ghash.c [hunk 26]
```diff
@@ -0,0 +1,883 @@
...
+  if(success) {
+    pmem_nvm_flush((void*)(entries + node_index), sizeof(paddr_t));
+  }
+  return success;
+    
+
+#ifndef SIMPLE_ENTRIES
+    // ent->index = new_index;
+    // ent->size = new_range;
+#endif
```
Comment (iangneal) [inline kernfs/lpmem_ghash.c]: Is SIMPLE_ENTRIES used anywhere?

File: kernfs/lpmem_ghash.h [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+#include "global/global.h"
+#include "shared.h"
+
+// from common
+#include "common/common.h"
+
+// local includes
+#include "hash_functions.h"
+
+#define SIMPLE_ENTRIES
```
Comment (iangneal) [inline kernfs/lpmem_ghash.h]: Is this used fore anything anymore? Can we consolidate?

File: kernfs/tests/mkfs.sh (hunk 18 above)
Comment (iangneal) [reply kernfs/tests/mkfs.sh:1]: Ideally I'd like to remove all sudos since it's easier to automate.

File: libfs/src/filesystem/fs.c [hunk 28]
```diff
@@ -1225,16 +1431,29 @@ ssize_t do_unaligned_read(struct inode *ip, uint8_t *dst, offset_t off, size_t i
...
           bmap_req.start_offset = off_aligned;
           bmap_req.blk_count = 1;
+          bmap_req_arr.start_offset = off_aligned;
+          bmap_req_arr.blk_count = 1;
           if (enable_perf_stats) {
               start_tsc = asm_rdtscp();
           }
-          ret = bmap(ip, &bmap_req);
+          if(IDXAPI_IS_HASHFS()) {
+            ret = bmap_hashfs(ip, &bmap_req_arr);
```
Comment (iangneal) [inline libfs/src/filesystem/fs.c:1426]: 4 space indent

File: libfs/src/filesystem/fs.c [hunk 29]
```diff
@@ -1225,16 +1431,29 @@ ssize_t do_unaligned_read(struct inode *ip, uint8_t *dst, offset_t off, size_t i
...
+            ret = bmap(ip, &bmap_req);
+          }
           if (enable_perf_stats) {
               g_perf_stats.tree_search_tsc += (asm_rdtscp() - start_tsc);
               g_perf_stats.tree_search_nr++;
           }
           mlfs_assert(ret != -EIO);
-          bh = bh_get_sync_IO(bmap_req.dev, bmap_req.block_no, BH_NO_DATA_ALLOC);
+          if(IDXAPI_IS_HASHFS()) {
+            bh = bh_get_sync_IO(bmap_req.dev, bmap_req_arr.block_no[0], BH_NO_DATA_ALLOC);
```
Comment (iangneal) [inline libfs/src/filesystem/fs.c:1437]: 4 space indent

File: libfs/src/filesystem/hash_functions.h [hunk 30]
```diff
@@ -0,0 +1 @@
+../../../../nvm-file-indexing/api/common/hash_functions.h
```
Comment (iangneal) [inline libfs/src/filesystem/hash_functions.h]: Still might want to copy this directly

File: libfs/tests/MTCCcorrect.c [hunk 31]
```diff
@@ -0,0 +1,260 @@
+#include <stdio.h>
```
Comment (iangneal) [inline libfs/tests/MTCCcorrect.c:3]: Comment at the top

File: libfs/tests/Makefile [hunk 32]
```diff
@@ -123,6 +125,12 @@ age: age.c
 
 MTCC: MTCC.c
 	$(CC) -g -o $@ $^  -I$(INCLUDES) -L$(LIBFS_DIR) -lmlfs -L$(LIBSPDK_DIR) -lspdk -DMLFS $(CFLAGS) $(LDFLAGS)
+MTCCcorrect: MTCCcorrect.c
+	$(CC) -g -o $@ $^  -I$(INCLUDES) -L$(LIBFS_DIR) -lmlfs -L$(LIBSPDK_DIR) -lspdk -DMLFS $(CFLAGS) $(LDFLAGS)
+write_bench: write_bench.c
```
Comment (iangneal) [inline libfs/tests/Makefile]: Remove the rules if the files are gone.

File: shim/glibc-2.19/intl/plural.c [hunk 33]
```diff
@@ -1,8 +1,8 @@
-/* A Bison parser, made by GNU Bison 2.7.  */
```
Comment (iangneal) [inline shim/glibc-2.19/intl/plural.c:1]: Remove this file if you can

**Review by iangneal — CHANGES_REQUESTED** [review]
Round 2

**Comment by iangneal** [comment]
I think all the comments remaining are just nitpicks, so just fix them when you can.

File: bench/BenchmarkProcesses.py (hunk 1 above)
Comment (eshiple) [reply bench/BenchmarkProcesses.py:56]: you might be right... but we use sudo for the mlfs.mkfs script normally. Or at least I have been.

File: kernfs/lpmem_cuckoohash.h (hunk 24 above)
Comment (eshiple) [reply kernfs/lpmem_cuckoohash.h]: no

File: kernfs/lpmem_cuckoohash.c (hunk 23 above)
Comment (eshiple) [reply kernfs/lpmem_cuckoohash.c]: no

File: kernfs/lpmem_cuckoo_hash_impl.h (hunk 22 above)
Comment (eshiple) [reply kernfs/lpmem_cuckoo_hash_impl.h]: no

File: kernfs/lpmem_cuckoo_hash_impl.c (hunk 21 above)
Comment (eshiple) [reply kernfs/lpmem_cuckoo_hash_impl.c]: no

**Review by iangneal — APPROVED** [review]

**Pull request merged**